src/test/java/com/ibm/disni/benchmarks/PingPongClient.java

```diff
@@ -0,0 +1,149 @@
+  RdmaActiveEndpointGroup<CustomClientEndpoint> endpointGroup;
+  private InetSocketAddress address;
+  private int bufferSize;
+  private int loopCount;
+
+  public CustomClientEndpoint createEndpoint(RdmaCmId idPriv, boolean serverSide) throws IOException {
+    return new CustomClientEndpoint(endpointGroup, idPriv, serverSide, bufferSize);
+  }
+
+  public void runTCP() throws Exception {
+    SocketChannel socketChannel = SocketChannel.open(address);
+    socketChannel.configureBlocking(true);
+    socketChannel.socket().setReceiveBufferSize(bufferSize);
+    socketChannel.socket().setSendBufferSize(bufferSize);
+    ByteBuffer sendBuf = ByteBuffer.allocateDirect(bufferSize);
+    ByteBuffer recvBuf = ByteBuffer.allocateDirect(bufferSize);
+    sendBuf.asCharBuffer().put("PING").clear();
+    long startTime = System.nanoTime();
+    for (int i = 0; i < loopCount; i++){
+      int read = 0;
+      int written = 0;
+
+      // Send PING
+      sendBuf.clear();
+      do {
+        written += socketChannel.write(sendBuf);
+      } while (written != bufferSize);
+
+      // Recv PONG
+      do {
+        read += socketChannel.read(recvBuf);
+      } while (read != bufferSize);
+      recvBuf.clear();
+    }
+    printResults(startTime);
+    socketChannel.close();
+  }
+
+  public void runRDMA() throws Exception {
+    //create a EndpointGroup. The RdmaActiveEndpointGroup contains CQ processing and delivers CQ event to the endpoint.dispatchCqEvent() method.
+    endpointGroup = new RdmaActiveEndpointGroup<CustomClientEndpoint>(1000, false, 128, 4, 128);
+    endpointGroup.init(this);
+    //we have passed our own endpoint factory to the group, therefore new endpoints will be of type CustomClientEndpoint
+    //let's create a new client endpoint
+    CustomClientEndpoint endpoint = endpointGroup.createEndpoint();
+
+    //connect to the server
+    endpoint.connect(address, 1000);
+    System.out.println("PingPongClient::client channel set up ");
+
+    //in our custom endpoints we have prepared (memory registration and work request creation) some memory
+    //buffers beforehand.
+    //let's send one of those buffers out using a send operation
+    ByteBuffer sendBuf = endpoint.getSendBuf();
+    ByteBuffer recvBuf = endpoint.getRecvBuf();
+    sendBuf.asCharBuffer().put("PING");
+    SVCPostSend postSend = endpoint.postSend(endpoint.getWrList_send());
+    SVCPostRecv postRecv = endpoint.postRecv(endpoint.getWrList_recv());
+    long startTime = System.nanoTime();
+    for (int i = 0; i < loopCount + 1; i++) {
+      // Send PING
+      sendBuf.clear();
+      postSend.execute();
+      endpoint.getWcEvents().take();
+
+      // Recv PONG
+      postRecv.execute();
+      endpoint.getWcEvents().take();
+      recvBuf.clear();
+    }
+
+    printResults(startTime);
+    //close everything
+    endpoint.close();
+    endpointGroup.close();
+	System.exit(0);
+  }
+
```

Comment:
Can you explain me why we need an extra PingPong benchmark? It seems it does pretty much what the SendRecv benchmark is doing already.

Reply:
SendRecv benchmark sends bulk of Send and Recv requests in parallel. PingPong sends one by one Send and Recv request (only when it got send it sends Recv). This makes 1:1 correspondence with TCP Socket benchmark.

Reply:
It sends `queueDepth` number of send requests at the beginning. To compare aples:aples with TCP need whether to run SendRecv benchmark with queue depth = 1 or have `queueDepth` TCP sockets to server or send `queueDepth * bufferSize` bigger buffer in TCP :
```
actualRecvs = 64
actualSends = 64
opCount 64
actualRecvs = 1
actualSends = 1
opCount 65
actualRecvs = 1
actualSends = 1
opCount 66
actualRecvs = 3
actualSends = 3
opCount 69
actualRecvs = 5
actualSends = 5
opCount 74
actualRecvs = 4
actualSends = 4
opCount 78
```
In PingPong benchmark flow state is controlled by benchmark itself:
```
send PING
recv PONG
send PING
...
```

Reply:
Also PingPong allows to test on big buffers (100MB+) where RDMA shows good performance compared to TCP. With queueDepth  need to allocate multiple such buffers.

Reply:
Sorry, still confused, when you say "It sends queueDepth number of send requests at the beginning." you mean the new PingPong benchmark does that?

Reply:
No PingPong is basically SendRecv with `queueDepth` = 1. Existing SendRecvClient has configuration parameter `queueDepth` which is 64 by default. This makes difficult to map to TCP world (64 connections or 64 bigger payload size, etc). To makes things simpler PingPong compares single qp with queueDepth = 1 with a single TCP socket pair. They both exchanges buffer of specified size. It doesn't designed to get full network bandwidth, just to rough estimate RDMA vs TCP latency on a single connection.

Reply:
The [line you pointed](https://github.com/zrlio/disni/blob/master/src/test/java/com/ibm/disni/benchmarks/SendRecvClient.java#L77):
```java
int actualSends = endpoint.sendRequest(actualRecvs); // actualRecvs = 64 at the beginning

public int sendRequest(int res) throws IOException {
  int ready = Math.min(res, sendBudget); // res=64, sendBudget = 64
  for (int i = 0; i < ready; i++){ 
    int index = sendIndex % pipelineLength;
    sendIndex++;
    sendCall[index].execute(); // 64 times it calls this sendCall
    sendBudget--;
    sendPending++;
  }
  return ready;
 }
```

Reply:
In PingPong `actualRecvs` and `actualSends` would be always 1

Reply:
Let me rephrase, my previous answer is misleading. The way SendRecv works from what I see is that if queuedepth is set to 1, it's doing exactly what PingPong. You can set queudepth to a larger value, then SendRecv will send a batch of requests at the beginning.

Reply:
Peter, I just had a look at the SendRecv benchmark code and with queue depth 1 it does exactly what your benchmark is supposed to do. The server initially waits for the client to send a request and sends on back once it received it from the client and so forth...

Reply:
Yes, the problem is that it's not fair to benchmark against TCP single socket with `queueDepth` > 1. That's why i've made separate benchmark to just compare RDMA vs TCP. Because user may forget to set `queueDepth` = 1 and compare single socket result with 64 queueDepth RDMA (even with 1 queueDepth RDMA showing very good result on 10Kb+ buffer sizes). Also to make iteration loop clearer to show that RDMA and TCP doing exactly the same thing (to not confuse with `actualRecv`, etc.).

Reply:
If that's the issue then we just change the default value from 64 to 1. I do agree that the SendRecv code is a bit difficult to understand, but I don't think we want two benchmarks in the code doing exactly the same in a different way. We could replace the existing SendRecv benchmark with the PingPong benchmark, but then we would loose functionality (the SendRecv benchmark can work with different queue depths).

Reply:
Ok, i'll add TCP part to SendRecv test. We'll tell our users that was interested in this functionality to run RDMA mode with `-q 1` when comparing to TCP

Reply:
@patrickstuedi sure. Just updated existing tests with TCP mode.